### DISCUSSION TITLE
Valetudo 2026.01.0

### DISCUSSION THREAD

**Discussion by Hypfer** [Hypfer/Valetudo#2402] [Release Discussions]
<div align="center">
    <img src="https://raw.githubusercontent.com/Hypfer/Valetudo/master/assets/logo/valetudo_logo_with_name.svg" width="800" alt="valetudo">
    <p align="center"><h2>2026.01.0</h2></p>
</div>

Starting fresh into the next demi-decade

# Valetudo 2026.01.0

Kicking off 2026, Valetudo now comes with improved visuals. Don't worry, it's not YouTube. You can stop screaming.
The soul is still here. I'm just iterating.

## A new font

Most central to this is the new choice of font. Replacing the default material Roboto, Valetudo now uses IBM Plex Sans.

I really like that font. It is mostly Helvetica but IBM and with some twists and interesting visual features.
Compared to Roboto, which is engineered to be in the background, IBM Plex Sans tells you that it exists and has something to show to you.

<img width="933" height="283" alt="image" src="https://github.com/user-attachments/assets/e8c44cb7-dcf6-4820-9bfd-62220177df0b" />

Beside the merits of the font itself, I think out of all corps, IBM is probably the most aligned with what Valetudo is.
When I think IBM, I think massive Mainframes, sturdy Thinkpads and lasting substance.
Serious business that didn't bother with petty nonsense like rebranding or new logos since the early 70s.

They are a corp and have skeletons in their basement for sure, but for the most part, they just do their thing, building their **I**nternational **B**usiness **M**achines and keeping the world running.
A refreshing contrast to the rest of big tech, currently at the forefront of dismantling democracy, turning everything into dysfunctional gambling, just being evil and generally setting things on fire.

Of course, I'm also writing this to invoke Cunningham's law and get someone to tell me about all the evil stuff IBM does, so that I can return to my cynical baseline.

Anyway.
I think this gives Valetudo a more distinct visual identity, which fits the maturity and spirit of the project.

## Improved icons

Being stuck sick at home and not attending 39C3, I still wanted to do something that isn't hacking vacuum robots for a change.
Fortunately, I have this whole project that comes with many different hats I can wear, so I've decided to wear the design hat and take a vacation from Valetudo by working on Valetudo.

The map view always felt a bit like it was built by an engineer; possibly because it was built by an engineer.
Therefore, I stared at it for a while and pondered _why_ that is. 🔮

After investigating at what the vendors do and what makes their visual language tick, I noticed that my icons were just.. bland and flat.
A unicolor gray circle on a gray background.

So the first thing I changed was to add more visual detail. Three circles are more interesting than one circle.
Additionally, I've noticed that the vendors use slight gradients to give the icons presence in the space instead of appearing like someone layered cardboard cutouts on top of something. So I did that too, and indeed it suddenly starts feeling much better.

<img width="1024" height="561" alt="image" src="https://github.com/user-attachments/assets/d81ab92a-c095-4697-b84c-e6608b592be9" />

Just by looking at the comparisons, I could not tell you that there is a gradient now. It just feels.. better.
So that is neat.
As it should turn out however, Firefox got _very unhappy_ rendering all those gradient-filled SVGs at 60fps.

I suppose that makes sense. Filling a simple shape with a single color is much less computationally expensive than calculating the color to fill with for each single pixel. It also doesn't help that the map renderer architecture is not at all following best practices on how to build performant 2d graphics.

So that came with a technical detour, adding automatic mipmap generation and letting the renderer use raster graphics instead, while still keeping the visual clarity of the SVGs.

While I was at it, I've also cleaned up some other performance issues. For example, also caused by the architecture, using shadows and blur on the canvas completely tanks performance due to.. idk I didn't care enough at this time. Something something very large things to draw.

Lastly, I've made the red and purple of virtual restrictions a bit less pure red and purple, as those didn't really fit the rest of the map.

## Improved visuals

Beside icons, other visuals were also improved.

For example, text on the map now comes in these pills:

<img width="472" height="421" alt="image" src="https://github.com/user-attachments/assets/ea2e4908-d985-4b7a-b8cd-dec722ab3320" />

This greatly elevates the look from "meme-font Impact with black and white stroke" to "how all the other vendor apps are looking", which isn't always a correct goal, but this time it sure is.

Additionally, handling of zoomlevels was improved. The text now scales more sensibly and so do the icons.
For some reason, past-me thought that icons must never get smaller than x pixels, leading to this abomination:

<img width="142" height="101" alt="image" src="https://github.com/user-attachments/assets/698ee398-77b0-425a-980a-fe7f80ad1c0a" />

Now, new year new me 💅 is doing the sensible thing:

<img width="128" height="108" alt="image" src="https://github.com/user-attachments/assets/c011c703-f3d7-45ea-bb66-98a5f62296aa" />

Don't ask me why it took me so long to realize that the previous thing was nonsense.

## Floor Materials

Not all is just visual in this update, though this also comes with new visuals.
After just going through backlogs of messages and generally looking around and assessing the situation, I've noticed that @xuefer apparently managed to draw floor materials in what I assume must be a private fork of Valetudo. (Which is great! Everyone should do it like that. It is FOSS precisely so that this can happen. People solving their own problems.)

This intrigued me, because I just never had an idea how to implement that. Seeing it being possible though made things finally click.
Once that mental block is gone, it is actually surprisingly simple. All you need to do is render some pixels in a different color and then you get texture.
So thank you @xuefer, for being a model citizen!

<img width="893" height="816" alt="image" src="https://github.com/user-attachments/assets/bfbcb07f-4933-4950-99b9-1ca770d45653" />

From a technical standpoint, this works by checking the coordinates of a pixel drawn, determining if according to the texture it should be accented, and if so, draw with an accented color. Simple stuff and super easy to extend with more textures.

So now, if your robot's firmware supports it, you can configure the floor material of a segment.

What is this good for, you wonder? That's where the next item comes in

## FloorMaterialDirectionAwareNavigationControlCapability

That's a mouthful.
If your firmware supports it, you can now tell the robot "these are wood planks and you will hurt them and possibly you if you don't clean them in paths aligned with them".

<img width="539" height="95" alt="image" src="https://github.com/user-attachments/assets/04246fe7-a93a-45c5-9d27-f319b2c7f42e" />

I don't have such flooring, nor would I want to have it the same way I do not need non-dishwasher-safe things in my life.
But if you do, this will likely improve your experience.

## What else is going on?

If you're reading this, you probably have a 3d printer, or are at the very least aware of their existence.
What you may not know however is that we by now have similar machines for similar prices for embroidery.

@jhbruhn recently got one, but was unhappy that it required an account and the vendor app to use.
So of course he did what we all should do when facing such tech: He liberated it through making it interoperable.

<img width="720" height="815" alt="image" src="https://github.com/user-attachments/assets/32ff40e2-cfe2-4c13-b2b3-52f589024db0" />

Using [https://github.com/jhbruhn/respira](https://github.com/jhbruhn/respira) you can now use the Brother PP-1 SKiTCH embroidery machine completely without the cloud.

If you or someone in your household might be interested in these crafts, now is the time to look into that.

I am unfortunately not a very artsy person, so I cannot really paint vivid pictures for you regarding what could all be done with that thing.
I am great at being cynical though!

So one thing you _could_ do is create counterfeits of work clothes, which may or may not allow someone with criminal energy to enter places they should not be in. Getting busted for being a darknet embroidery vendor will also most certainly get you onto some podcasts. Of course, please don't actually do that. Just because you don't see them behind your screen, doesn't mean that cybercrime would have no victims.

Putting that aside, I do think that there is real potential for ethical red-teaming here. You could easily manufacture the right clothing on-site in the trunk of your car. That _also_ makes for a good story and gets you onto podcasts, but it does so by improving the world instead of harming it.

Maybe we as the hacker bubble could push for a bit more ethics and not reward criminals as much as we tend to sometimes do.
Just some food for thought.

I will leave you with this perfectly cut clip from 39C3 I found on the internet (needs sound):

https://github.com/user-attachments/assets/f783baea-c897-426f-9ccb-e5a673a0bd10

Happy new year!

## The usual

If you like this release or Valetudo in general, you might want to consider donating:

[https://github.com/sponsors/Hypfer](https://github.com/sponsors/Hypfer)

[https://builder.dontvacuum.me/donations.txt](https://builder.dontvacuum.me/donations.txt)

### Features

- **ui**: Update capability icons with new monochrome robot icon design [`5004619`](http://github.com/Hypfer/Valetudo/commit/5004619db24c2002dc98164f44685a246b4ef034)
- **ui**: Improve some wording [`d6cb9c3`](http://github.com/Hypfer/Valetudo/commit/d6cb9c39dfac311f5c2538c2da81cf5c136d3e46)
- **webserver**: Lawful compliance [`6638b23`](http://github.com/Hypfer/Valetudo/commit/6638b23883879bab99f6f34f36f460ccb765c5d2)
- **ui**: Update robot icons to look less like a pokeball [`386dbd8`](http://github.com/Hypfer/Valetudo/commit/386dbd8eb382135bcbe12e77449833a80aff87d1)
- **ui**: Switch font to IBM Plex Sans [`d0fa5a2`](http://github.com/Hypfer/Valetudo/commit/d0fa5a210821851672a8e3383ca09325cfceb0cd)
- **ui**: Improve label rendering for map structures [`8fd17be`](http://github.com/Hypfer/Valetudo/commit/8fd17bede7c4ce72a2cfc8dc4a1d4356a5bda21b)
- **ui**: General map polish including improvements to performance + UX + looks [`6d54afa`](http://github.com/Hypfer/Valetudo/commit/6d54afa02077aecaff958ad61b9a679c0d1144f5)
- **ui**: Improve map icons [`05450c4`](http://github.com/Hypfer/Valetudo/commit/05450c4e5a6606ea97df8e6d6bdea8d7b57c6d98)
- **vendor.midea**: More quirks [`8dfb08c`](http://github.com/Hypfer/Valetudo/commit/8dfb08cbfcea600ae405f2536911362f1d23cd56)
- **vendor.dreame**: FloorMaterialDirectionAwareNavigationControlCapability [`2c4717f`](http://github.com/Hypfer/Valetudo/commit/2c4717fe6ba5432fa4c6cd98b7dae22550481d88)
- **vendor.roborock**: FloorMaterialDirectionAwareNavigationControlCapability + MapSegmentMaterialControlCapability [`5c257a5`](http://github.com/Hypfer/Valetudo/commit/5c257a52c86db9a05a6c6210793e47fa8154ef58)
- **ui**: FloorMaterialDirectionAwareNavigationControlCapability [`4005e0e`](http://github.com/Hypfer/Valetudo/commit/4005e0e65e418ab7ef3e233ab4f6a6095556555f)
- **core**: FloorMaterialDirectionAwareNavigationControlCapability [`7db5b75`](http://github.com/Hypfer/Valetudo/commit/7db5b756bba6b95e8712d27d510e170103af2403)
- **webserver**: Add another 404 page [`3456363`](http://github.com/Hypfer/Valetudo/commit/345636319d8dde0d5ba15b30c0c464a6065ba846)
- **vendor.dreame**: MapSegmentMaterialControlCapability [`5273df1`](http://github.com/Hypfer/Valetudo/commit/5273df12c6de7eca9d4e6f9f8d6065d8748f23f3)
- **vendor.midea**: MapSegmentMaterialControlCapability [`93132c6`](http://github.com/Hypfer/Valetudo/commit/93132c66adca976050df6f262b5e567a6cae736f)
- **ui**: MapSegmentMaterialControlCapability [`b13d796`](http://github.com/Hypfer/Valetudo/commit/b13d7964565f929e57fc43e97ad18b760f1f8f50)
- **core**: MapSegmentMaterialControlCapability [`6a8dd6f`](http://github.com/Hypfer/Valetudo/commit/6a8dd6f5bb23f750532ea05570b9801650f974f0)

### Fixes

- **ui**: Minor visual cleanup [`4cbc004`](http://github.com/Hypfer/Valetudo/commit/4cbc0041766002a4c3362eae3373ca0317df30c4)
- **vendor.dreame**: The latest L10spuh firmware uses the newer-style mop extension controls [`04bac60`](http://github.com/Hypfer/Valetudo/commit/04bac60fdc988db2074c3a13638379715591f6f6)

### Refactoring

- **vendor.roborock**: Wording [`714af6c`](http://github.com/Hypfer/Valetudo/commit/714af6c41b61cb2a280411b1144ccf2d4451c7a2)
- **ui**: Clean up mapLayer color handling [`2146713`](http://github.com/Hypfer/Valetudo/commit/21467136e8756a2ec2419001ad115ef57c37e3a4)

### Chores

- **release**: 2026.01.0 [`4323938`](http://github.com/Hypfer/Valetudo/commit/43239380164236b77c9ad2fc85f50948b502193e)
- Update eslint rules [`dde428f`](http://github.com/Hypfer/Valetudo/commit/dde428fe381d27b7f9187b86157c109e973e5f27)

<hr /><em>This discussion was created from the release <a href='https://github.com/Hypfer/Valetudo/releases/tag/2026.01.0'>Valetudo 2026.01.0</a>.</em>

**Comment by Nekkowe** [comment]
Oooh, this is great! Love the new icons, especially the charging dock & the more obvious red "delete wall (etc.)" one. 

The floor materials look great, too! I haven't quite figured out how to get those to show - maybe I have to let it recreate the whole map once, or maybe the Roborock S5 just doesn't support detecting them.  

I don't think preventing the zone labels from getting too small was that bad of an idea, not at all - after the change, on some of my devices, they've become kind of unreadable unless I zoom waaaay in, even on the default zoom level:
<img width="320" height="491" alt="image" src="https://github.com/user-attachments/assets/81892251-8e8d-445c-8cb4-88de4637b09d" />
But that's not the end of the world, either, since you don't have to click/tap the triangle dead-on to select the zone and the names don't matter all that much.

Lastly, something I'm realizing I really appreciate: even if you zoom waaaay out on purpose (like to fit some kind of massive office building, maybe) the robot's own icon has *great* contrast against the rest of the map - even at this tiny scale, you can always tell exactly where it is! That's neat.

<img width="46" height="68" alt="image" src="https://github.com/user-attachments/assets/48b74082-2e14-4b18-aef3-e30b9893d149" />

Thank you so much for this project!
Get well soon, and happy new year! :partying_face: 

(I had no idea about machine embroidery… hearing that you can just freely draw up designs in Inkscape has me *really* tempted, that's so cool…!)

**Comment by lyonzy** [comment]
As an IBMer (opinions my own etc) that's such a refreshing comment to read! And it's correct too IMO, we do things like build [LLMs](https://www.ibm.com/granite) based on fully transparent training data and [ditched facial recognition technology over law enforcement concerns](https://www.theguardian.com/technology/2020/jun/09/ibm-quits-facial-recognition-market-over-law-enforcement-concerns). I'm sure there are skeletons somewhere but overall integrity is prioritised and it shows.
